# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGL2ContextTransformFeedback.cpp
# Commit: ccfa767dba64
# Full Hash: ccfa767dba644dc193e0246eb8e8ff3377e8b8a5
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Client-side bindings mirror for precise CC, and merge similar codepaths. r=handyman
#   
#   * Context loss using RAII
#   * Move Program reflection Client-side
#   
# ==============================================================================

diff -r 71c122ac0ca7 -r ccfa767dba64 dom/canvas/WebGL2ContextTransformFeedback.cpp
--- a/dom/canvas/WebGL2ContextTransformFeedback.cpp	Wed Jan 08 22:19:14 2020 +0000
+++ b/dom/canvas/WebGL2ContextTransformFeedback.cpp	Wed Jan 08 22:19:16 2020 +0000
@@ -4,7 +4,6 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 #include "WebGL2Context.h"
-#include "WebGLActiveInfo.h"
 #include "WebGLProgram.h"
 #include "WebGLTransformFeedback.h"
 #include "GLContext.h"
@@ -14,51 +13,20 @@
 // -------------------------------------------------------------------------
 // Transform Feedback
 
-already_AddRefed<WebGLTransformFeedback>
-WebGL2Context::CreateTransformFeedback() {
+RefPtr<WebGLTransformFeedback> WebGL2Context::CreateTransformFeedback() {
   const FuncScope funcScope(*this, "createTransformFeedback");
   if (IsContextLost()) return nullptr;
 
   GLuint tf = 0;
   gl->fGenTransformFeedbacks(1, &tf);
 
-  RefPtr<WebGLTransformFeedback> ret = new WebGLTransformFeedback(this, tf);
-  return ret.forget();
+  return new WebGLTransformFeedback(this, tf);
 }
 
-void WebGL2Context::DeleteTransformFeedback(WebGLTransformFeedback* tf) {
-  const FuncScope funcScope(*this, "deleteTransformFeedback");
-  if (!ValidateDeleteObject(tf)) return;
-
-  if (tf->mIsActive) {
-    ErrorInvalidOperation("Cannot delete active transform feedbacks.");
-    return;
-  }
-
-  if (mBoundTransformFeedback == tf) {
-    BindTransformFeedback(LOCAL_GL_TRANSFORM_FEEDBACK, nullptr);
-  }
-
-  tf->RequestDelete();
-}
-
-bool WebGL2Context::IsTransformFeedback(
-    const WebGLTransformFeedback* const obj) {
-  const FuncScope funcScope(*this, "isTransformFeedback");
-  if (!ValidateIsObject(obj)) return false;
-
-  if (obj->IsDeleteRequested()) return false;
-
-  return obj->mHasBeenBound;
-}
-
-void WebGL2Context::BindTransformFeedback(GLenum target,
-                                          WebGLTransformFeedback* tf) {
-  const FuncScope funcScope(*this, "bindTransformFeedback");
+void WebGL2Context::BindTransformFeedback(WebGLTransformFeedback* tf) {
+  FuncScope funcScope(*this, "bindTransformFeedback");
   if (IsContextLost()) return;
-
-  if (target != LOCAL_GL_TRANSFORM_FEEDBACK)
-    return ErrorInvalidEnum("`target` must be TRANSFORM_FEEDBACK.");
+  funcScope.mBindFailureGuard = true;
 
   if (tf && !ValidateObject("tf", *tf)) return;
 
@@ -72,13 +40,16 @@
 
   ////
 
-  mBoundTransformFeedback = (tf ? tf : mDefaultTransformFeedback);
+  mBoundTransformFeedback = (tf ? tf : mDefaultTransformFeedback.get());
 
-  gl->fBindTransformFeedback(target, mBoundTransformFeedback->mGLName);
+  gl->fBindTransformFeedback(LOCAL_GL_TRANSFORM_FEEDBACK,
+                             mBoundTransformFeedback->mGLName);
 
   if (mBoundTransformFeedback) {
     mBoundTransformFeedback->mHasBeenBound = true;
   }
+
+  funcScope.mBindFailureGuard = false;
 }
 
 void WebGL2Context::BeginTransformFeedback(GLenum primMode) {
@@ -110,8 +81,8 @@
 }
 
 void WebGL2Context::TransformFeedbackVaryings(
-    WebGLProgram& program, const nsTArray<nsString>& varyings,
-    GLenum bufferMode) {
+    WebGLProgram& program, const std::vector<std::string>& varyings,
+    GLenum bufferMode) const {
   const FuncScope funcScope(*this, "transformFeedbackVaryings");
   if (IsContextLost()) return;
 
@@ -120,14 +91,4 @@
   program.TransformFeedbackVaryings(varyings, bufferMode);
 }
 
-Maybe<WebGLActiveInfo> WebGL2Context::GetTransformFeedbackVarying(
-    const WebGLProgram& program, GLuint index) {
-  const FuncScope funcScope(*this, "getTransformFeedbackVarying");
-  if (IsContextLost()) return Nothing();
-
-  if (!ValidateObject("program", program)) return Nothing();
-
-  return program.GetTransformFeedbackVarying(index);
-}
-
 }  // namespace mozilla